In [1]:
import pandas as pd
import numpy as np

In [2]:
source_dir = '../data/undesired_outcomes_pred/'
dest_dir = source_dir
extension = '.csv'

In [3]:
elog = pd.read_csv(source_dir.__add__('elog.csv'))

In [4]:
pd.options.display.max_columns = 100

In [5]:
elog.sample(7)

Case ID                                       Activity  \
1620033  9ee5e14f6ab5434e      Payment application-Application-calculate   
1313202  19a310079e7580b8                  Reference alignment-Main-save   
1132568  861a41247e2fa3c0    Payment application-Application-mail income   
124438   a7f1ec20307e594b                         Inspection-Remote-plan   
1697941  69d45362d1eb012a         Geo parcel document-Reported-calculate   
361810   3397e8da41191736                Parcel document-Main-initialize   
1891008  660d8c5aee0cd7bc  Geo parcel document-Declared-finish pre-check   

                              Resource       Complete Timestamp  \
1620033                         03b214  2016-12-14 15:23:57.081   
1313202                         7fb8a5  2016-11-22 13:32:18.000   
1132568                            NaN  2016-05-13 00:00:00.000   
124438            Inspection automaton  2015-07-19 15:51:21.848   
1697941                         727350  2017-11-10 21:53:07.175   
361810   Document processing automaton  2015-06-11 07:35:12.000   
1891008  Document processing automaton  2017-05-24 06:18:30.251   

         (case) amount_applied0  (case) amount_applied1  \
1620033                36918.56                     NaN   
1313202                10065.27                     NaN   
1132568                 7161.68                     NaN   
124438                  1438.85                     NaN   
1697941                 4688.47                     NaN   
361810                  6024.42                     NaN   
1891008                 3928.57                     NaN   

         (case) amount_applied2  (case) amount_applied3  (case) applicant  \
1620033                     NaN                     NaN  4c25fed3bef0f945   
1313202                     NaN                     NaN  54d3be56c394240b   
1132568                     NaN                     NaN  9c488ae83909c71a   
124438                      NaN                     NaN  a897608d24926af0   
1697941                     NaN                     NaN  134efb4b60ab7ea8   
361810                      NaN                     NaN  cee00c603f722570   
1891008                     NaN                     NaN  3adb3efbeb7ac64f   

         (case) area  (case) basic payment  (case) cross_compliance  \
1620033     125.1964                  True                      0.0   
1313202      29.7041                  True                      0.0   
1132568      20.8915                  True                      0.0   
124438        4.1348                  True                      0.0   
1697941      13.9695                  True                      0.0   
361810       18.2269                  True                      0.0   
1891008      12.3107                  True                      0.0   

        (case) department  (case) greening  (case) number_parcels  \
1620033                e7             True                     56   
1313202                e7             True                     11   
1132568                6b             True                      9   
124438                 d4             True                      3   
1697941                6b             True                      8   
361810                 6b             True                     12   
1891008                4e             True                     11   

         (case) payment_actual0  (case) payment_actual1  \
1620033                36933.37                     NaN   
1313202                10068.68                     NaN   
1132568                 7149.54                     NaN   
124438                  1438.85                     NaN   
1697941                 4687.62                     NaN   
361810                  5992.14                     NaN   
1891008                 3927.95                     NaN   

         (case) payment_actual2  (case) payment_actual3  (case) penalty_ABP  \
1620033                     NaN                     NaN               False   
1313202                     N

In [6]:
elog.loc[~pd.isnull(elog['(case) penalty_amount2'])]

Case ID                                           Activity  \
8383     b0d199e3fd21ffb4        Payment application-Application-mail income   
8384     b0d199e3fd21ffb4            Entitlement application-Main-mail valid   
8385     b0d199e3fd21ffb4         Payment application-Application-mail valid   
8386     b0d199e3fd21ffb4            Entitlement application-Main-mail valid   
8387     b0d199e3fd21ffb4                    Control summary-Main-initialize   
8388     b0d199e3fd21ffb4                 Control summary-Main-begin editing   
8389     b0d199e3fd21ffb4                Control summary-Main-finish editing   
8390     b0d199e3fd21ffb4                    Parcel document-Main-initialize   
8391     b0d199e3fd21ffb4                 Parcel document-Main-begin editing   
8392     b0d199e3fd21ffb4               Parcel document-Main-insert document   
8393     b0d199e3fd21ffb4                          Parcel document-Main-save   
8394     b0d199e3fd21ffb4                Parcel document-Main-finish editing   
8395     b0d199e3fd21ffb4                Reference alignment-Main-initialize   
8396     b0d199e3fd21ffb4                 Reference alignment-Main-performed   
8397     b0d199e3fd21ffb4          Department control parcels-Main-performed   
8398     b0d199e3fd21ffb4      Department control parcels-Main-begin editing   
8399     b0d199e3fd21ffb4            Entitlement application-Main-initialize   
8400     b0d199e3fd21ffb4                      Reference alignment-Main-save   
8401     b0d199e3fd21ffb4                          Parcel document-Main-save   
8402     b0d199e3fd21ffb4                Parcel document-Main-finish editing   
8403     b0d199e3fd21ffb4         Entitlement application-Main-begin editing   
8404     b0d199e3fd21ffb4             Entitlement application-Main-calculate   
8405     b0d199e3fd21ffb4      Department control parcels-Main-begin editing   
8406     b0d199e3fd21ffb4                      Reference alignment-Main-save   
8407     b0d199e3fd21ffb4      Department control parcels-Main-begin editing   
8408     b0d199e3fd21ffb4         Payment application-Application-initialize   
8409     b0d199e3fd21ffb4      Payment application-Application-begin editing   
8410     b0d199e3fd21ffb4          Payment application-Application-calculate   
8411     b0d199e3fd21ffb4             Entitlement application-Main-calculate   
8412     b0d199e3fd21ffb4             Entitlement application-Main-calculate   
...                   ...                                                ...   
1641626  f24f9e1ac0d7701f               Payment application-Objection-refuse   
1641627  f24f9e1ac0d7701f  Payment application-Objection-begin editing fr...   
1641628  f24f9e1ac0d7701f            Payment application-Objection-calculate   
1641629  f24f9e1ac0d7701f            Payment application-Objection-calculate   
1641630  f24f9e1ac0d7701f  Payment application-Objection-begin editing fr...   
1641631  f24f9e1ac0d7701f  Payment application-Objection-begin editing fr...   
1641632  f24f9e1ac0d7701f            Payment application-Objection-calculate   
1641633  f24f9e1ac0d7701f  Payment application-Objection-begin editing fr...   
1641634  f24f9e1ac0d7701f            Payment application-Objection-calculate   
1641635  f24f9e1ac0d7701f  Payment application-Objection-begin editing fr...   
1641636  f24f9e1ac0d7701f      Payment application-Objection-revoke approval   
1641637  f24f9e1ac0d7701f  Payment application-Objection-begin admissibil...   
1641638  f24f9e1ac0d7701f  Payment application-Objection-check admissibility   
1641639  f24f9e1ac0d7701f            Payment application-Objection-calculate   
1641640  f24f9e1ac0d7701f               Payment application-Objection-refuse   
1641641  f24f9e1ac0d7701f  Payment application-Objection-begin editing fr...   
1641642  f24f9e1ac0d7701f      Payment application-Objection-insert document   
1641643  f24f9e1ac0d7701f                      Payment application-Main-save   
1641644  f24f9e1ac

In [7]:
elog.loc[~pd.isnull(elog['(case) payment_actual3'])].sample(3)

Case ID                                           Activity  \
34025    31dc792faebd46d5  Payment application-Objection-check admissibility   
1495161  6dcac1562d4ca766                            Inspection-On-Site-save   
1561706  57e032f39476cb61        Payment application-Objection-begin payment   

        Resource       Complete Timestamp  (case) amount_applied0  \
34025     79367e  2017-11-03 12:01:37.251               150960.97   
1495161   7850e8  2016-12-01 12:19:24.231                34769.12   
1561706     DP-Z  2017-01-19 10:25:43.582                49143.90   

         (case) amount_applied1  (case) amount_applied2  \
34025                      0.00               150927.32   
1495161                34794.71                35254.05   
1561706                49143.90                49106.81   

         (case) amount_applied3  (case) applicant  (case) area  \
34025                 150959.87  e5a229e3a5f5915f     398.9333   
1495161                35253.72  a5d52351e3091a8d     107.9741   
1561706                49106.81  0a93de7f894be319     164.3606   

         (case) basic payment  (case) cross_compliance (case) department  \
34025                    True                      0.0                6b   
1495161                  True                      0.0                d4   
1561706                  True                      0.0                4e   

         (case) greening  (case) number_parcels  (case) payment_actual0  \
34025               True                     96               150537.94   
1495161             True                     59                29492.92   
1561706             True                     35                42445.17   

         (case) payment_actual1  (case) payment_actual2  \
34025                      0.00                 -789.85   
1495161                 5273.58                 -376.22   
1561706                 6718.39                 -118.20   

         (case) payment_actual3  (case) penalty_ABP  (case) penalty_AGP  \
34025                    786.37               False               False   
1495161                 -378.30               False               False   
1561706                 -841.24                True                True   

         (case) penalty_AJLP  (case) penalty_AUVP  (case) penalty_AVBP  \
34025                  False                False                False   
1495161                False                False                False   
1561706                False                 True                False   

         (case) penalty_AVGP  (case) penalty_AVJLP  (case) penalty_AVUVP  \
34025                  False                 False                 False   
1495161                False                 False                 False   
1561706                False                 False                 False   

         (case) penalty_B16  (case) penalty_B2  (case) penalty_B3  \
34025                 False               True              False   
1495161               False              False              False   
1561706               False              False              False   

         (case) penalty_B4  (case) penalty_B5  (case) penalty_B5F  \
34025                False              False               False   
1495161              False               True               False   
1561706              False              False               False   

         (case) penalty_B6  (case) penalty_BGK  (case) penalty_BGKV  \
34025                False               False                False   
1495161              False               False                False   
1561706              False               False                False   

         (case) penalty_BGP  (case) penalty_C16  (case) penalty_C4  \
34025                 False               False              False   
1495161               False               False              False   
1561706               False               False              False   

         (case) penalty_C9  (case) penalty_CC  (case

### Case lifecycle

In [8]:
len(set(elog['Case ID']))

43809

So there are 43809 cases present in the event log. Let's see how many of these cases are actually useful to use for training/testing/cross-validation.

In [9]:
elog.loc[elog['Activity'] == 'Payment application-Application-decide'].shape

(65437, 89)

In [10]:
len(set(elog.loc[elog['Activity'] == 'Payment application-Application-decide']['Case ID']))

43793

We see that there are only 43809 - 43793 = 16 cases that do not have the the 'Payment application-Application-decide' activity in their lifecycle (as far as can be deduced from the event logs)

Let's drop these 16 cases since they are incomplete in the context of this business question and are not that significant when one considers the number of cases at hand.

In [11]:
to_be_dropped_cases = set(elog['Case ID']).difference(set(set(elog.loc[elog['Activity'] == 'Payment application-Application-decide']['Case ID'])))

In [12]:
elog = elog[~elog['Case ID'].isin(to_be_dropped_cases)]

In [13]:
len(set(elog['Case ID'])) # for confirmation

43793

But since we want to be able to detect the undesired outcomes before the 'Payment application-Application-decide' the choice is made to consider the event that proceeds this activity in almost all the cases: 'Payment application-Application-finish editing'.

In [14]:
len(set(elog.loc[elog['Activity'] == 'Payment application-Application-finish editing']['Case ID']))

43793

In [15]:
elog_uo = elog.loc[elog['Activity'] == 'Payment application-Application-finish editing']

In [16]:
len(set(elog_uo['Case ID']))

43793

In [18]:
elog_uo.drop_duplicates(subset='Case ID', keep='first', inplace=True)

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [21]:
elog_uo.shape[0] == len(set(elog_uo['Case ID']))

True

### Undesired outcome #1

In [22]:
set(elog['Activity'])

{'Control summary-Main-begin editing',
 'Control summary-Main-change department',
 'Control summary-Main-finish editing',
 'Control summary-Main-initialize',
 'Control summary-Main-insert document',
 'Control summary-Main-save',
 'Control summary-Main-take original document',
 'Department control parcels-Main-begin editing',
 'Department control parcels-Main-change department',
 'Department control parcels-Main-insert document',
 'Department control parcels-Main-performed',
 'Department control parcels-Main-remove document',
 'Department control parcels-Main-save',
 'Entitlement application-Change-begin editing',
 'Entitlement application-Change-calculate',
 'Entitlement application-Change-change department',
 'Entitlement application-Change-decide',
 'Entitlement application-Change-finish editing',
 'Entitlement application-Change-initialize',
 'Entitlement application-Change-insert document',
 'Entitlement application-Change-remove document',
 'Entitlement application-Change-revoke d

In [23]:
def is_undesired_outcome1(x):
    case_id = x['Case ID']
    case_subset = elog[elog['Case ID'] == case_id]
    case_events = set(case_subset['Activity'])
    if 'Payment application-Application-begin payment' in case_events: # there has been a begin payment activity
        if 'Payment application-Application-abort payment' not in case_events: # no abort payment activity in the data
            return True
        else: # check whether it was before the end of the year
            earliest_date = min(list(case_subset['timestamp_year']))
            date_abort_payment = list(case_subset[case_subset['Activity'] == 
                                             'Payment application-Application-abort payment']['timestamp_year'])[0]
            if date_abort_payment > earliest_date:
                return True
    return False

In [ ]:
elog_uo['undesired_outcome1'] = elog_uo.apply(is_undesired_outcome1, axis=1)

In [ ]:
print("shape ", elog_uo.shape[0])
print("true " , elog_uo[elog_uo['undesired_outcome1'] == True].shape[0])
print("false " , elog_uo[elog_uo['undesired_outcome1'] == False].shape[0])

### Undesired outcome #2

In [ ]:
def is_undesired_outcome2(x):
    case_id = x['Case ID']
    case_subset = elog[elog['Case ID'] == case_id]
    case_subprocesses = set(case_subset['subprocess'])
    if 'Change' in case_subprocesses or 'Objection' in case_subprocesses:
        return True
    else:
        return False

In [ ]:
elog_uo['undesired_outcome2'] = elog_uo.apply(is_undesired_outcome2, axis=1)

In [ ]:
elog_uo['undesired_outcome'] = np.where(elog_uo['undesired_outcome1'] | elog_uo['undesired_outcome2'], True, False)

In [ ]:
print("shape ", elog_uo.shape[0])
print("true " , elog_uo[elog_uo['undesired_outcome2'] == True].shape[0])
print("false " , elog_uo[elog_uo['undesired_outcome2'] == False].shape[0])

In [ ]:
elog_uo.to_csv(dest_dir.__add__('elog_uo').__add__(extension), index=False)